Embed API Client: JavaScript

sgammon opened the pull request:
This changeset adds a prepackaged API client in JavaScript, starting with the *Embed API*, which allows Embedded Menus to be placed on some webpage or other display. This client is low-level and offers advanced access to embeddable configuration details.

Client code is dynamically configured via JSON discovery documents.

Commit:
Embed API Client: JavaScript

This changeset adds a prepackaged API client in JavaScript, starting
with the *Embed API*, which allows Embedded Menus to be placed on some
webpage or other display. This client is low-level and offers advanced
access to embeddable configuration details.

Client code is dynamically configured via JSON discovery documents.

sgammon commented on JavaScript/Makefile:
```diff
@@ -0,0 +1,68 @@
+## Makefile: Bloombox API Client for JavaScript
+#
+
+TARGET ?= target
```
full-featured `Makefile`

tsamartino replied:
So long tho

sgammon commented on JavaScript/bower.json:
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "Bloombox API Client",
```
eventually it'll support being installed via `bower`

tsamartino replied:
Sweeet

sgammon commented on JavaScript/gulpfile.js:
```diff
@@ -0,0 +1,249 @@
+         sources: []
+           .concat(sources.scripts.common)
+           .concat(sources.scripts.api.embed)
+           .concat(sources.scripts.api.all)}]}};
```
modular closure support via `gulp`

sgammon commented on JavaScript/karma.conf.js:
```diff
@@ -0,0 +1,40 @@
+    autoWatchBatchDelay: 1000,
+    frameworks: ['mocha', 'chai'],
+    reporters: ['dots', 'progress', 'spec', 'coverage'],
+    browsers: ["ChromeCanary"],
```
browser testing configuration

sgammon commented on JavaScript/src/common/api.js:
```diff
@@ -0,0 +1,128 @@
+ * @param {?Object=} opt_payload Method payload body to send.
+ * @constructor
+ */
+function RPC(api, method, props, opt_payload) {
```
RPC layer from `common`

sgammon commented on JavaScript/src/api/embed/index.js:
```diff
@@ -0,0 +1,144 @@
+ *
+ * @const {string}
+ */
+EmbedAPI.serviceName = "embed";
```
support for the Embed API

sgammon commented on JavaScript/src/api/tools.js:
```diff
@@ -0,0 +1,62 @@
+  /* istanbul ignore if  */
+  if ("gapi" in window && "client" in window["gapi"]) {
+    return callback();
+  }
```
support for loading the API client into the browser context

Commit:
Small code updates, Bower support

Commit:
Ignore DS_Store files

tsamartino commented on JavaScript/Bloombox JavaScript Client.ipr:
```diff
@@ -0,0 +1,656 @@
+        <option name="myIgnoreVendorSpecificProperties" value="false" />
+      </inspection_tool>
+      <inspection_tool class="XmlWrongFileType" enabled="false" level="WARNING" enabled_by_default="false" />
+      <inspection_tool class="unused" enabled="true" level="WARNING" enabled_by
```
How did you decide which inspection tools to enable/disable?

sgammon replied:
i do this from intelliJ, and there is a standard set attached to each of our polymer components in `.eslint` files.

basically, we enable all the Closure stuff, and disable any incompatible stuff (we're in the browser, so no NodeJS, and so on). then, some of it is up to style - for instance, we use 120-character lines with 2-space indentation.

sgammon replied:
FYI there are more tools that run in CI, so these are just the static analysis engines we run *while* developing - since we're commenting on the IDEA project config file ;)

tsamartino commented on JavaScript/Makefile:
```diff
@@ -0,0 +1,73 @@
+	@echo
+	@echo "Bloombox API: JavaScript Client"
+	@echo
+	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
```
This command seems fun

sgammon replied:
yes, this one is from a blog post. it basically just generates a help section for the `Makefile`, so you can run `make help`. Give it a shot ;)

I include it with everything I build now. it builds the help file simply by parsing for the `target:  ## comment` pattern, extracting `comment` to be the help text for `target`, using `awk` and `grep` which are standard Unix tools.

tsamartino commented on JavaScript/bower.json:
```diff
@@ -0,0 +1,29 @@
+    "tests"
+  ],
+  "dependencies": {
+    "polymer": "Polymer/polymer#^1.7.0"
```
Older version of polymer, updates are available

sgammon replied:
Good catch, will update 👍

tsamartino commented on JavaScript/gulpfile.js:
```diff
@@ -0,0 +1,249 @@
+gulp.task('pre-test', function () {
+  return gulp.src(['src/**/*.js'])
+    // Covering files
+    .pipe(istanbul())
```
Does Istanbul export oil via these pipes?

sgammon replied:
🌞  🐫  🇹🇷  🛢

sgammon replied:
no but really, istanbul is for code coverage - it's a JS package, and we're piping it thru our javascript files here so it can catch the files it needs to measure, before we run our tests.

tsamartino commented on JavaScript/src/api/all.js:
```diff
@@ -0,0 +1,52 @@
+  Logging['say']('Bloombox:API', 'Loading discovery documents...',
+      window['BloomboxAPI']['list']);
+
+  var promises = [];
```
*List full of empty promises*

sgammon replied:
*Glass half full: List of fulfilled promises*

tsamartino commented on JavaScript/src/api/embed/index.js:
```diff
@@ -0,0 +1,155 @@
+ * @return {Promise} self, for chaining.
+ */
+EmbeddedMenu.prototype.data = (function(cbk, err) {
+  'use strict';
```
Is this ready to go using live menu data or is dummy data from dev still being used?

sgammon replied:
it's live data 🌠 

basically, all of how that works is in _Feature: Embedded Menus_ (momentum/bloombox#328). it's a rather complex mechanism to sum up here, but our embedded clients don't even talk to Firebase directly for the menu data. they only talk to it to check the *version* of the menu data, to save Firebase load/bandwidth.

then, the client downloads that version of the data from static, pre-prepared JSON on Cloud Storage. this greatly reduces our attack surface for DDoS (especially if the attacker is ourselves, due to embedded web menus bringing in the sum total of our _customers'_ traffic) and allows us to serve it via CDN, which is faster.

because the JSON is static, we can also cache it infinitely - so if the menu doesn't change, the client can re-use their copy.

all of these are measures designed to both:
1) reduce cost, especially from our riskiest sources - i.e. losing Firebase is disastrous
2) increase performance, especially with caching - which usually costs someone else

tsamartino commented on JavaScript/test/testsuite.js:
```diff
@@ -0,0 +1,180 @@
+    Logging["severe"]("Test", "Hello from a test.");
+    severe("Test", "Hello from a test.");
+  });
+});
```
Do the above error messages need updating from filler strings?

sgammon replied:
nah - notice the filename. this is _actually_ a test file, haha. this line is actually testing logging - this lib comes with 60% coverage out of the box, although the tools to report that on the Pull Request aren't up yet

tsamartino reviewed (commented):
I realize this isn't the codebase that would contain it, but do we check our database against a JSON schema anywhere so that junk stuff doesn't end up in it? Or do we use a different method for data validation?

sgammon commented:
to answer your main question, @tsamartino, for now, this SDK is read-only. so there's no need to validate data because we never accept any data, besides their API key, partner ID, location ID, and the "menu style," which is a string value defaulting-to and only supporting the value `MASTER_ONLY`, for now.

you are right to point that out, though, and it's not a bad idea to ship some JSON schema files with the SDK that enable validation and modeling on the client-side for things like strains, product flags, and dispensaries.

tsamartino commented:
Writing JSON schemas is totally doable, probably would have prevented a lot of errors early on with crap data somehow making its way into firebase. Only thing I haven't actually figured out is the best way to check data against that schema. My current flow involves a lot of copying and pasting to a validator tool available online lol. 

All this looks good to go for me, thanks for answering those questions. 

Sent from my iPhone

> On Apr 26, 2017, at 10:24 AM, Sam Gammon <notifications@github.com> wrote:
> 
> @sgammon commented on this pull request.
> 
> In JavaScript/test/testsuite.js:
> 
> > +    Logging["warn"]("Test", "Hello from a test.");
> +    warn("Test", "Hello from a test.");
> +  });
> +
> +  it('should expose "error"', function() {
> +    expect(Logging["error"]).to.not.equal(undefined);
> +    Logging["error"]("Test", "Hello from a test.");
> +    error("Test", "Hello from a test.");
> +  });
> +
> +  it('should expose "severe"', function() {
> +    expect(Logging["severe"]).to.not.equal(undefined);
> +    Logging["severe"]("Test", "Hello from a test.");
> +    severe("Test", "Hello from a test.");
> +  });
> +});
> nah - notice the filename. this is actually a test file, haha. this line is actually testing logging - this lib comes with 60% coverage out of the box, although the tools to report that on the Pull Request aren't up yet
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

Commit:
Early support for Polymer/Closure, small refactors

- Centralize exports and script boot
- Isolate script boot so as not to force inclusion
- Create minimal copy of requisite API code for Polymer

Commit:
Fixes to build chain

tsamartino commented:
(Approve)

Commit:
Add routine to build JS release tarball

Commit:
Ignore JS release tarballs

Commit:
Use NPM on Buildbot

Commit:
Add gulp-rename to lockfile

Commit:
Add del module to lockfile

Commit:
Setup headless browser testing in CI

Commit:
Updates to testing, gulpfile, Polymer support

Commit:
Add READMEs for main repo and Polymer
